### PR TITLE
storage: add InitialHeartBeatFailedError as a structured error

### LINKED ISSUE #27109: storage: initial connection heartbeat failed error gets propagated as an ambiguous error
While investigating a client issue we noticed that in some clusters when a node went down, other live nodes in the cluster were unable to evict stale entries from their range descriptor caches leading to underreplicated ranges and stuck queries. 

What we think happened is the following:

- We have a multinode cluster.
- Some nodes' cached range descriptors only contain one replica for a range.
- A node (N1) goes down
- The nodes which only had range descriptor cache entries for N1 for certain ranges, now try to update their node liveness.
- The nodes can't connect to N1 because it's down resulting in an [initial connection heartbeat failed error in context.go](https://github.com/cockroachdb/cockroach/blob/a4296f8cab552dfaa174cb9d03aa46cf78336385/pkg/storage/node_liveness.go#L713).

- Upstream in dist sender the error is [not treated as a RequestDidNotStart error](https://github.com/cockroachdb/cockroach/blob/77fb319b0e10524798bb9b49388b7a879be11040/pkg/kv/dist_sender.go#L1327) and the transaction also contains a commit, so it's treated as an ambiguous error.
- We don't evict entries from the range descriptor cache for ambiguous errors, only send error, so now we don't get fresh information for ranges that that only have N1 cache entries.

### PR DESCRIPTION
**Pull request by m-schneider**
This patch makes the GRPC Connect function return a structured error
when it's unable to connect and has never seen a heartbeat on that
connection. Before it returned an ambiguous error which prevented the
range descrptor cache from evicting stale entries upstream. In some cases
this could lead to a node being unable to find nodes that have replicas
for a given range because its cache is stale, but it doesn't refresh it
because there's an ambiguous error.

Closes #27109

Release note: None

### CODE REVIEW

**Review by bdarnell — COMMENTED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27295)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/roachpb/errors.proto, line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27295#-LH-VPxA5UVCIZNwHOhO:-LH-VPxA5UVCIZNwHOhP:b-7wb8xn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a253672ada5cbae12c2d99aa2e04f1fd8191fb06/pkg/roachpb/errors.proto#L249)):*
> ```protobuf
> // connection to a node, we aren't successful and have never seen a
> // heartbeat over that connection before.
> message InitialHeartBeatFailedError {
> ```

All our other protobuf errors are members of the ErrorDetail union. And it's slightly lossy to wrap the inner error in a roachpb.Error. So it doesn't feel quite right to make this a protobuf error, but it also doesn't seem right to put a non-protobuf error in roachpb. 

One more suggestion for a place to put this error: pkg/util/netutil. It's only tangentially related, but it's not too bad and is low enough level to avoid dependency cycles (I think - if I'm missing something and it won't work there, then just leave it in roachpb)

**Comment by m-schneider** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/roachpb/errors.proto, line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27295#-LH-VPxA5UVCIZNwHOhO:-LH-gUnvBEBQh25-BO-K:bqypqtv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a253672ada5cbae12c2d99aa2e04f1fd8191fb06/pkg/roachpb/errors.proto#L249)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

All our other protobuf errors are members of the ErrorDetail union. And it's slightly lossy to wrap the inner error in a roachpb.Error. So it doesn't feel quite right to make this a protobuf error, but it also doesn't seem right to put a non-protobuf error in roachpb. 

One more suggestion for a place to put this error: pkg/util/netutil. It's only tangentially related, but it's not too bad and is low enough level to avoid dependency cycles (I think - if I'm missing something and it won't work there, then just leave it in roachpb)
</blockquote></details>

That makes sense. I moved the error into netutil and also kept the inner error as an error.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27295)*

**Review by bdarnell — APPROVED** [review]

**Review by nvb — APPROVED** [review]
Nice!

File: pkg/util/netutil/net.go [hunk 1]
```diff
@@ -159,3 +160,21 @@ func FatalIfUnexpected(err error) {
 		log.Fatal(context.TODO(), err)
 	}
 }
+
+// InitialHeartBeatFailedError indicates that while attempting a GRPC
+// connection to a node, we aren't successful and have never seen a
+// heartbeat over that connection before.
+type InitialHeartBeatFailedError struct {
```
Comment (nvanbenschoten) [inline pkg/util/netutil/net.go]: nit: "Heartbeat" is one camel-cased word everywhere else in our codebase. Let's keep it the same here.

**Comment by m-schneider** [comment]
___
*[pkg/util/netutil/net.go, line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27295#-LH-gBM8CYQ6B8lYTo4--r2-167:-LH-pgJG6jHnqYbE_9wH:b-80tlqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/598caf4a6ad88e7ecacd42b7612e26d80edb936b/pkg/util/netutil/net.go#L167)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: "Heartbeat" is one camel-cased word everywhere else in our codebase. Let's keep it the same here.
</blockquote></details>

Sure!

**Comment by tschottdorf** [comment]
LGTM. It would've been nice to test that this actually works, though. Any `errors.Wrap(err, "something")` would foul up this mechanism. (I verified manually that there isn't any wrapping of the error between its creation in the transport and consumption in DistSender, but that's the kind of thing that *could* break silently in the future).

**Pull request merged**